src/Simulation/QCTraceSimulator/SortedQubitPool.cs

```diff
@@ -0,0 +1,329 @@
+        /// Depth times are assumed to be precise for comparison.
+        /// </summary>
+        /// <param name="time">Time to subtract from this object.</param>
+        /// <returns>Result of subtraction.</returns>
+        internal ComplexTime Subtract(ComplexTime time) {
+            if (DepthTime == time.DepthTime) {
+                return new ComplexTime(0, TrailingZeroDepthGateCount - time.TrailingZeroDepthGateCount);
+            }
+            double result = DepthTime - time.DepthTime;
+            if (result <= 0) {
+                // This could only happen due to insufficient floating point calculation precision.
+                throw new ArgumentException("Result of ComplexTime subtraction is not positive.");
+            }
+            return new ComplexTime(result, 0);
+        }
+
+        /// <summary>
+        /// Compares two complex times. First DepthTime is compared, then TrailingZeroDepthGateCount is compared.
+        /// </summary>
+        /// <returns>0 when a = b, -1 when a&lt; b, and 1 when a &gt; b</returns>
+        internal static int Compare(ComplexTime a, ComplexTime b) {
+            int result = a.DepthTime.CompareTo(b.DepthTime);
+            if (result != 0) {
+                return result;
+            }
+            return a.TrailingZeroDepthGateCount.CompareTo(b.TrailingZeroDepthGateCount);
+        }
+
+        /// <summary>
+        /// Finds smallest of the two ComplexTime arguments.
+        /// </summary>
+        /// <returns>Smallest of the two arguments according to comparison.</returns>
+        internal static ComplexTime Min(ComplexTime a, ComplexTime b) {
+            return Compare(a, b) < 0 ? a : b;
+        }
+
+        /// <summary>
+        /// Finds largest of the two ComplexTime arguments.
+        /// </summary>
+        /// <returns>Largest of the two arguments according to comparison.</returns>
+        internal static ComplexTime Max(ComplexTime a, ComplexTime b) {
+            return Compare(a, b) > 0 ? a : b;
+        }
+
+        /// <summary>
+        /// Minimal value of ComplexTime.
+        /// </summary>
+        internal static ComplexTime MinValue { get; } = new ComplexTime(double.MinValue, long.MinValue);
+
+        /// <summary>
+        /// Maximum value of ComplexTime.
+        /// </summary>
+        internal static ComplexTime MaxValue { get; } = new ComplexTime(double.MaxValue, long.MaxValue);
+
+        /// <summary>
+        /// Zero ComplexTime.
+        /// </summary>
+        internal static ComplexTime Zero { get; } = new ComplexTime(0.0, 0);
+    }
+
+    /// <summary>
+    /// Stores qubits sorted by corresponding ComplexTime values. For example, a ComplexTime value that correponds to a qubit
+    /// could be the qubit availability time: the end time of the last gate on the qubit.
+    /// Supports queries to retrieve qubit with maximum time less than given and minimum time greater than given Sample.
```

Comment:
It might be useful to specify the time complexity (from perf point of view) of the supported operations on the pool.

Reply:
I did specify time complexity on the methods themselves. This whole class is here to have them at O( log(N) ).